File: src/com/timepath/hl2/swing/VBFCanvas.java
package com.timepath.hl2.swing;

import com.timepath.hl2.io.VTF;
import com.timepath.hl2.io.font.VBF;
import com.timepath.hl2.io.font.VBF.BitmapGlyph;
import java.awt.*;
import java.awt.event.*;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;

/**
 *
 * @author TimePath
 */
public class VBFCanvas extends JPanel implements MouseListener, MouseMotionListener {

    private static final Logger LOG = Logger.getLogger(VBFCanvas.class.getName());

    private static final AlphaComposite acNormal = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 1);

    private static final AlphaComposite acSelected = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.5f);

    private static final AlphaComposite acText = AlphaComposite.getInstance(AlphaComposite.SRC_OVER);

    private static final long serialVersionUID = 1L;

    private Image img;

    private Point last;

    private List<BitmapGlyph> selected = new LinkedList<BitmapGlyph>();

    private VBF vbf;

    private VTF vtf;

    /**
     * Creates new form VBFTest
     */
    public VBFCanvas() {
        this.addMouseListener(this);
        this.addMouseMotionListener(this);
    }

    public List<BitmapGlyph> get(Point p) {
        List<BitmapGlyph> intersected = new LinkedList<BitmapGlyph>();
        if(vbf == null || vbf.getGlyphs() == null) {
            return intersected;
        }
        for(BitmapGlyph g : vbf.getGlyphs()) {
            if(g.getBounds().contains(p)) {
                intersected.add(g);
            }
        }
        return intersected;
    }

    @Override
    public Dimension getPreferredSize() {
        if(vbf == null) {
            return new Dimension(128, 128);
        }
        return new Dimension(vbf.getWidth(), vbf.getHeight());
    }

    public BitmapGlyph getSelected() {
        if(this.selected.isEmpty()) {
            return null;
        }
        return this.selected.get(0);
    }

    public void setVBF(VBF b) {
        this.vbf = b;
        this.revalidate();
    }

    public void setVTF(VTF t) {
        this.vtf = t;
        this.vbf.setWidth(vtf.getWidth());
        this.vbf.setHeight(vtf.getHeight());
        this.repaint();
    }

    public void mouseClicked(MouseEvent me) {
    }

    public void mouseDragged(MouseEvent me) {
        Point p = me.getPoint();
        if(last != null) {
            me.translatePoint(-last.x, -last.y);
        }
        for(BitmapGlyph sel : selected) {
            if(SwingUtilities.isRightMouseButton(me)) {
                sel.getBounds().width += me.getPoint().x;
                sel.getBounds().height += me.getPoint().y;
            } else {
                sel.getBounds().x += me.getPoint().x;
                sel.getBounds().y += me.getPoint().y;
            }
        }
        last = p;
        this.repaint();
    }

    public void mouseEntered(MouseEvent me) {
    }

    public void mouseExited(MouseEvent me) {
    }

    public void mouseMoved(MouseEvent me) {
    }

    public void mousePressed(MouseEvent me) {
        last = me.getPoint();
        if(!SwingUtilities.isLeftMouseButton(me)) {
            return;
        }
        List<BitmapGlyph> old = new LinkedList<BitmapGlyph>(selected);
        List<BitmapGlyph> clicked = get(me.getPoint());
        for(BitmapGlyph g : clicked) {
            if(selected.contains(g)) {
                return;
            }
        }
        if(me.isControlDown()) {
            selected.addAll(clicked);
        } else {
            selected = clicked;
        }
        for(BitmapGlyph g : old) {
            this.repaint(g.getBounds());
        }
        for(BitmapGlyph g : selected) {
            this.repaint(g.getBounds());
        }
    }

    public void mouseReleased(MouseEvent me) {
    }

    public void select(BitmapGlyph g) {
        selected.clear();
        if(g != null) {
            selected.add(g);
            this.repaint(g.getBounds());
        }
    }

    @Override
    protected void paintComponent(Graphics graphics) {
        Graphics2D g = (Graphics2D) graphics;
        g.setComposite(acNormal);
        g.setColor(Color.BLACK);
        g.fillRect(0, 0, this.getWidth(), this.getHeight());
        if(img == null && vtf != null) {
            try {
                img = vtf.getImage(vtf.getMipCount() - 1);
            } catch(IOException ex) {
                LOG.log(Level.SEVERE, null, ex);
            }
        }
        if(this.vbf != null) {
            g.setColor(Color.GRAY);
            g.fillRect(0, 0, vbf.getWidth(), vbf.getHeight());
        }
        if(img != null) {
            g.drawImage(img, 0, 0, this);
        }
        if(this.vbf != null) {
            for(BitmapGlyph glyph : vbf.getGlyphs()) {
                if(glyph == null) {
                    continue;
                }
                Rectangle bounds = glyph.getBounds();
                if(bounds == null || bounds.isEmpty()) {
                    continue;
                }
                g.setComposite(acNormal);
                g.setColor(Color.GREEN);
                g.drawRect(bounds.x, bounds.y, bounds.width - 1, bounds.height - 1);
                if(selected.contains(glyph)) {
                    g.setComposite(acSelected);
                    g.fillRect(bounds.x, bounds.y, bounds.width - 1, bounds.height - 1);
                }
//                 TODO: Negative font folor
//                Map<TextAttribute, Object> map = new Hashtable<TextAttribute, Object>();
//                map.put(TextAttribute.SWAP_COLORS, TextAttribute.SWAP_COLORS_ON);
//                map.put(TextAttribute.FOREGROUND, Color.BLACK);
//                map.put(TextAttribute.BACKGROUND, Color.TRANSLUCENT);
//                Font f = this.getFont().deriveFont(map);
//                g.setFont(f);
//                g.setXORMode(Color.WHITE);
                g.setComposite(acText);
                g.setColor(Color.GREEN);
                g.drawString(Integer.toString(glyph.getIndex()), bounds.x + 1, bounds.y + bounds.height - 1);
            }
        }
    }

}

File: src/main/kotlin/com/timepath/hl2/swing/VBFCanvas.kt
package com.timepath.hl2.swing

import com.timepath.hl2.io.font.VBF
import com.timepath.hl2.io.image.VTF
import java.awt.*
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import java.awt.event.MouseMotionListener
import java.io.IOException
import java.util.LinkedList
import java.util.logging.Level
import java.util.logging.Logger
import javax.swing.JPanel
import javax.swing.SwingUtilities

/**
 * @author TimePath
 */
public class VBFCanvas
/**
 * Creates new form VBFTest
 */
: JPanel(), MouseListener, MouseMotionListener {
    private var img: Image? = null
    private var last: Point? = null
    var selected: MutableList<VBF.BitmapGlyph> = LinkedList()
    private var vbf: VBF? = null
    private var vtf: VTF? = null

    init {
        addMouseListener(this)
        addMouseMotionListener(this)
    }

    public fun setVBF(vbf: VBF) {
        this.vbf = vbf
        revalidate()
    }

    public fun setVTF(t: VTF) {
        vtf = t
        vbf!!.width = vtf!!.width.toShort()
        vbf!!.height = vtf!!.height.toShort()
        repaint()
    }

    override fun mouseClicked(e: MouseEvent) {
    }

    override fun mousePressed(e: MouseEvent) {
        last = e.getPoint()
        if (!SwingUtilities.isLeftMouseButton(e)) {
            return
        }
        val old = LinkedList(selected)
        val clicked = get(e.getPoint())
        for (g in clicked) {
            if (g in selected) {
                return
            }
        }
        if (e.isControlDown()) {
            selected.addAll(clicked)
        } else {
            selected = clicked
        }
        for (g in old) {
            repaint(g.bounds)
        }
        for (g in selected) {
            repaint(g.bounds)
        }
    }

    fun get(p: Point): MutableList<VBF.BitmapGlyph> {
        val intersected = LinkedList<VBF.BitmapGlyph>()
        vbf?.let {
            for (g in it.glyphs) {
                if (p in g.bounds) {
                    intersected.add(g)
                }
            }
        }
        return intersected
    }

    override fun mouseReleased(e: MouseEvent) {
    }

    override fun mouseEntered(e: MouseEvent) {
    }

    override fun mouseExited(e: MouseEvent) {
    }

    override fun mouseDragged(e: MouseEvent) {
        val p = e.getPoint()
        last?.let {
            e.translatePoint(-it.x, -it.y)
        }
        for (sel in selected) {
            if (SwingUtilities.isRightMouseButton(e)) {
                sel.bounds.width += e.getPoint().x
                sel.bounds.height += e.getPoint().y
            } else {
                sel.bounds.x += e.getPoint().x
                sel.bounds.y += e.getPoint().y
            }
        }
        last = p
        repaint()
    }

    override fun mouseMoved(e: MouseEvent) {
    }

    public fun select(g: VBF.BitmapGlyph?) {
        selected.clear()
        g?.let { g ->
            selected.add(g)
            repaint(g.bounds)
        }
    }

    override fun paintComponent(g: Graphics) {
        val g2 = g as Graphics2D
        g2.setComposite(acNormal)
        g2.setColor(Color.BLACK)
        g2.fillRect(0, 0, getWidth(), getHeight())
        if ((img == null) && (vtf != null)) {
            try {
                img = vtf!!.getImage(0)
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, null, ex)
            }

        }
        vbf?.let {
            g2.setColor(Color.GRAY)
            g2.fillRect(0, 0, it.width.toInt(), it.height.toInt())
        }
        img?.let {
            g2.drawImage(it, 0, 0, this)
        }
        vbf?.let {
            for (glyph in it.glyphs) {
                if (glyph == null) {
                    continue
                }
                val bounds = glyph.bounds
                if ((bounds == null) || bounds.isEmpty()) {
                    continue
                }
                g2.setComposite(acNormal)
                g2.setColor(Color.GREEN)
                g2.drawRect(bounds.x, bounds.y, bounds.width - 1, bounds.height - 1)
                if (glyph in selected) {
                    g2.setComposite(acSelected)
                    g2.fillRect(bounds.x, bounds.y, bounds.width - 1, bounds.height - 1)
                }
                //                 TODO: Negative font folor
                //                Map<TextAttribute, Object> map = new Hashtable<TextAttribute, Object>();
                //                map.put(TextAttribute.SWAP_COLORS, TextAttribute.SWAP_COLORS_ON);
                //                map.put(TextAttribute.FOREGROUND, Color.BLACK);
                //                map.put(TextAttribute.BACKGROUND, Color.TRANSLUCENT);
                //                Font f = this.getFont().deriveFont(map);
                //                g.setFont(f);
                //                g.setXORMode(Color.WHITE);
                g2.setComposite(acText)
                g2.setColor(Color.GREEN)
                g2.drawString(Integer.toString(glyph.index.toInt()), bounds.x + 1, (bounds.y + bounds.height) - 1)
            }
        }
    }

    override fun getPreferredSize() = vbf?.let {
        Dimension(it.width.toInt(), it.height.toInt())
    } ?: Dimension(128, 128)

    companion object {

        private val LOG = Logger.getLogger(javaClass<VBFCanvas>().getName())
        private val serialVersionUID = 1
        private val acNormal = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 1f)
        private val acSelected = AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.5f)
        private val acText = AlphaComposite.getInstance(AlphaComposite.SRC_OVER)
    }
}
